File: lang/c/src/vtfppc/HOT.cpp
#include <vtfppc/HOT.h>

#include <vtfpp/HOT.h>

#include <sourceppc/Helpers.h>
#include <vtfppc/Convert.hpp>

using namespace sourceppc;
using namespace vtfpp;

SOURCEPP_API vtfpp_hot_handle_t vtfpp_hot_create() {
	return new HOT{};
}

SOURCEPP_API vtfpp_hot_handle_t vtfpp_hot_open_from_mem(const unsigned char* hotData, size_t hotLen) {
	SOURCEPP_EARLY_RETURN_VAL(hotData, nullptr);
	SOURCEPP_EARLY_RETURN_VAL(hotLen, nullptr);

	return new HOT{{reinterpret_cast<const std::byte*>(hotData), hotLen}};
}

SOURCEPP_API vtfpp_hot_handle_t vtfpp_hot_open_from_file(const char* hotPath) {
	SOURCEPP_EARLY_RETURN_VAL(hotPath, nullptr);

	return new HOT{std::string{hotPath}};
}

SOURCEPP_API int vtfpp_hot_is_valid(vtfpp_hot_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);

	return Convert::hot(handle)->operator bool();
}

SOURCEPP_API uint8_t vtfpp_hot_get_version(vtfpp_hot_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return Convert::hot(handle)->getVersion();
}

SOURCEPP_API void vtfpp_hot_set_version(vtfpp_hot_handle_t handle, uint8_t version) {
	SOURCEPP_EARLY_RETURN(handle);

	Convert::hot(handle)->setVersion(version);
}

SOURCEPP_API uint8_t vtfpp_hot_get_flags(vtfpp_hot_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return Convert::hot(handle)->getFlags();
}

SOURCEPP_API void vtfpp_hot_set_flags(vtfpp_hot_handle_t handle, uint8_t flags) {
	SOURCEPP_EARLY_RETURN(handle);

	Convert::hot(handle)->setFlags(flags);
}

SOURCEPP_API uint16_t vtfpp_hot_get_rect_count(vtfpp_hot_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return Convert::hot(handle)->getRects().size();
}

SOURCEPP_API vtfpp_hot_rect_t vtfpp_hot_get_rect_at_index(vtfpp_hot_handle_t handle, uint16_t index) {
	SOURCEPP_EARLY_RETURN_VAL(handle, VTFPP_HOT_RECT_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(Convert::hot(handle)->getRects().size() >= index, VTFPP_HOT_RECT_INVALID);

	const auto& rect = Convert::hot(handle)->getRects().at(index);
	return {
		.flags = rect.flags,
		.x1 = rect.x1,
		.y1 = rect.y1,
		.x2 = rect.x2,
		.y2 = rect.y2,
	};
}

SOURCEPP_API void vtfpp_hot_set_rect_at_index(vtfpp_hot_handle_t handle, uint16_t index, vtfpp_hot_rect_t rect) {
	SOURCEPP_EARLY_RETURN(handle);
	SOURCEPP_EARLY_RETURN(Convert::hot(handle)->getRects().size() >= index);

	auto& rect_ = Convert::hot(handle)->getRects().at(index);
	rect_.flags = static_cast<HOT::Rect::Flags>(rect.flags);
	rect_.x1 = rect.x1;
	rect_.y1 = rect.y1;
	rect_.x2 = rect.x2;
	rect_.y2 = rect.y2;
}

SOURCEPP_API void vtfpp_hot_add_rect(vtfpp_hot_handle_t handle, vtfpp_hot_rect_t rect) {
	SOURCEPP_EARLY_RETURN(handle);

	auto& rect_ = Convert::hot(handle)->getRects().emplace_back();
	rect_.flags = static_cast<HOT::Rect::Flags>(rect.flags);
	rect_.x1 = rect.x1;
	rect_.y1 = rect.y1;
	rect_.x2 = rect.x2;
	rect_.y2 = rect.y2;
}

SOURCEPP_API void vtfpp_hot_remove_rect_at_index(vtfpp_hot_handle_t handle, uint16_t index) {
	SOURCEPP_EARLY_RETURN(handle);
	SOURCEPP_EARLY_RETURN(Convert::hot(handle)->getRects().size() >= index);

	Convert::hot(handle)->getRects().erase(Convert::hot(handle)->getRects().begin() + index);
}

SOURCEPP_API sourcepp_buffer_t vtfpp_hot_bake(vtfpp_hot_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(Convert::hot(handle)->bake());
}

SOURCEPP_API int vtfpp_hot_bake_to_file(vtfpp_hot_handle_t handle, const char* hotPath) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(hotPath, false);

	return Convert::hot(handle)->bake(hotPath);
}

SOURCEPP_API void vtfpp_hot_close(vtfpp_hot_handle_t* handle) {
	SOURCEPP_EARLY_RETURN(handle);

	delete Convert::hot(*handle);
	*handle = nullptr;
}

File: lang/csharp/src/vtfpp/HOT.cs
using System;
using System.Runtime.InteropServices;

namespace sourcepp.vtfpp;

public sealed class HOT : sourcepp.ManagedNativeHandle
{
	[StructLayout(LayoutKind.Sequential, Pack = 1)]
	public struct Rect
	{
		[Flags]
		public enum Flags : byte
		{
			NONE              = 0,
			RANDOM_ROTATION   = 1 << 0,
			RANDOM_REFLECTION = 1 << 1,
			IS_ALTERNATE      = 1 << 2,
		}

		public Flags RectFlags;
		public ushort X1;
		public ushort Y1;
		public ushort X2;
		public ushort Y2;
	}

	internal HOT(nint handle) : base(handle, DLL.vtfpp_hot_close)
	{
	}

	public HOT() : this(DLL.vtfpp_hot_create())
	{
	}

	public HOT(ReadOnlySpan<byte> hotData) : this(DLL.vtfpp_hot_open_from_mem(hotData, (ulong) hotData.Length))
	{
	}
	
	public HOT(string hotPath) : this(DLL.vtfpp_hot_open_from_file(hotPath))
	{
	}
	
	public static implicit operator bool(HOT hot)
	{
		hot.ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_hot_is_valid(hot.Handle));
	}

	public byte Version
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_hot_get_version(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vtfpp_hot_set_version(Handle, value);
		}
	}

	public byte Flags
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_hot_get_flags(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vtfpp_hot_set_flags(Handle, value);
		}
	}

	public Rect[] Rects
	{
		get
		{
			ThrowIfDisposed();
			var rects = new Rect[DLL.vtfpp_hot_get_rect_count(Handle)];
			for (var i = 0; i < rects.Length; i++)
			{
				rects[i] = DLL.vtfpp_hot_get_rect_at_index(Handle, (ushort) i);
			}
			return rects;
		}
		set
		{
			ThrowIfDisposed();
			var storedRectCount = DLL.vtfpp_hot_get_rect_count(Handle);
			for (var i = 0; i < value.Length; i++)
			{
				if (i < storedRectCount)
				{
					DLL.vtfpp_hot_set_rect_at_index(Handle, (ushort) i, value[i]);
				}
				else
				{
					DLL.vtfpp_hot_add_rect(Handle, value[i]);
				}
			}
			for (var i = storedRectCount - value.Length - 1; i >= 0; i--)
			{
				DLL.vtfpp_hot_remove_rect_at_index(Handle, (ushort) i);
			}
		}
	}

	public byte[] Bake()
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_hot_bake(Handle)).Read<byte>();
	}

	public bool Bake(string hotPath)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_hot_bake_to_file(Handle, hotPath));
	}
}
